File: app/src/main/java/com/hop/pirate/model/impl/MineMachineListModelImpl.java
package com.hop.pirate.model.impl;

import android.content.Context;
import android.text.TextUtils;

import com.hop.pirate.PError;
import com.hop.pirate.R;
import com.hop.pirate.base.BaseModel;
import com.hop.pirate.callback.ResultCallBack;
import com.hop.pirate.greendao.MinerDaoUtil;
import com.hop.pirate.model.MineMachineListModel;
import com.hop.pirate.model.bean.MinerBean;
import com.hop.pirate.service.HopService;
import com.hop.pirate.service.SysConf;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

import androidLib.AndroidLib;
import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

/**
 * @description:
 * @author: mr.x
 * @date :   2020/5/30 2:10 PM
 */
public class MineMachineListModelImpl extends BaseModel implements MineMachineListModel {
    @Override
    public void getMinemachine(final Context context, final String address, final int random, final ResultCallBack<List<MinerBean>> resultCallBack) {
        schedulers(Observable.create(new ObservableOnSubscribe<List<MinerBean>>() {
            @Override
            public void subscribe(ObservableEmitter<List<MinerBean>> emitter) throws Exception {
                List<MinerBean> minerBeans;
                String miners = AndroidLib.randomMiner(address, random);
                if (miners.equals("null") || TextUtils.isEmpty(miners)) {
                    emitter.onError(new PError(context.getString(R.string.get_data_failed)));
                    return;
                }
                minerBeans = new ArrayList<>();
                JSONArray minerArr = new JSONArray(miners);
                for (int i = 0; i < minerArr.length(); i++) {
                    JSONObject obj = (JSONObject) minerArr.get(i);
                    MinerBean bean = new MinerBean();
                    bean.setMID(obj.optString("SubAddr"));
                    bean.setZone(obj.optString("Zone"));
                    bean.setMinerPoolAdd(SysConf.CurPoolAddress);
                    minerBeans.add(bean);
                }
                emitter.onNext(minerBeans);
                emitter.onComplete();
            }
        })).subscribe(new Observer<List<MinerBean>>() {
            @Override
            public void onSubscribe(Disposable d) {
                addSubscribe(d);
            }

            @Override
            public void onNext(List<MinerBean> walletBean) {
                resultCallBack.onSuccess(walletBean);
            }

            @Override
            public void onError(Throwable e) {
                resultCallBack.onError(e);
            }

            @Override
            public void onComplete() {
                resultCallBack.onComplete();
            }
        });
    }

    @Override
    public void removeAllSubscribe() {
        removeAllDisposable();
    }
}

File: app/src/main/java/com/hop/pirate/model/impl/MineMachineListModelImpl.kt
package com.hop.pirate.model.impl

import android.content.Context
import android.text.TextUtils
import androidLib.AndroidLib
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.hop.pirate.Constants
import com.hop.pirate.PirateException
import com.hop.pirate.R
import com.hop.pirate.base.BaseModel
import com.hop.pirate.callback.ResultCallBack
import com.hop.pirate.model.MineMachineListModel
import com.hop.pirate.model.bean.MinerBean
import com.hop.pirate.model.bean.OwnPool
import com.hop.pirate.model.bean.WalletBean
import com.hop.pirate.service.SysConf
import io.reactivex.Observable
import io.reactivex.ObservableOnSubscribe
import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.*

/**
 * @description:
 * @author: mr.x
 * @date :   2020/5/30 2:10 PM
 */
class MineMachineListModelImpl {
    suspend fun getMineMachine(address: String, random: Int): List<MinerBean> {
        return withTimeout(Constants.TIME_OUT.toLong()) {
            withContext(Dispatchers.IO) {
                val minersMachineStr = AndroidLib.randomMiner(address, random.toLong())
                val type = object : TypeToken<ArrayList<MinerBean>>() {}.type
                Gson().fromJson<ArrayList<MinerBean>>(minersMachineStr, type)
            }
        }
    }

    suspend fun ping(address: String): Double {
        return withTimeout(Constants.TIME_OUT.toLong()) {
            withContext(Dispatchers.IO) {
                val pingJson = AndroidLib.testPing(address)
                if (TextUtils.isEmpty(pingJson)) {
                    return@withContext -1.0
                } else {
                    val obj = JSONObject(pingJson)
                    obj.optDouble("ping")
                }
            }
        }

    }

}